File: frontend/src/api/progress.ts
import { api } from './client'

import type { Paper } from './syllabus'

/** Days to an exam, and how many of them are actually study days. */
export type ExamCountdown = {
  date: string
  days: number
  study_days: number
}

export type Countdown = {
  date: string
  prelims: ExamCountdown
  mains: ExamCountdown
}

export type BurndownPoint = {
  date: string
  /** Null in the future: the actual line stops at today. */
  remaining: number | null
  required: number
}

export type Burndown = {
  date: string
  total_leaves: number
  started_leaves: number
  remaining: number
  study_days_remaining: number
  required_per_day: number
  actual_per_day: number | null
  actual_window_days: number
  projected_finish: string | null
  series: BurndownPoint[]
}

export type PaperCoverage = {
  paper: Paper
  label: string
  leaves: number
  read: number
  revised: number
  tested: number
}

export type Coverage = {
  date: string
  papers: PaperCoverage[]
  totals: PaperCoverage | null
}

export type HeatmapCell = {
  node_id: string
  title: string
  paper: Paper
  section: string
  /** 1–5, or null for a topic never graded — an empty square, not a weak one. */
  confidence: number | null
  started: boolean
  next_due: string | null
  days_overdue: number
}

export type HeatmapSection = {
  paper: Paper
  label: string
  section: string
  cells: HeatmapCell[]
}

export type Heatmap = { date: string; sections: HeatmapSection[] }

export type EffortDay = {
  date: string
  minutes: number
  logs: number
  /** A planned off-day is marked, never drawn as a zero. */
  off: boolean
}

export type Effort = {
  date: string
  days: EffortDay[]
  total_minutes: number
  average_minutes: number
  study_days: number
}

export const getCountdown = () => api<Countdown>('/progress/countdown')

export const getBurndown = () => api<Burndown>('/progress/burndown')

export const getCoverage = () => api<Coverage>('/progress/coverage')

export const getHeatmap = (paper?: Paper) =>
  api<Heatmap>(`/progress/heatmap${paper ? `?paper=${paper}` : ''}`)

export const getEffort = (days = 30) =>
  api<Effort>(`/progress/streakless-summary?days=${days}`)

/** "6h 20m" — hours over a day's study, minutes below it. */
export function formatMinutes(minutes: number): string {
  if (minutes < 60) return `${minutes}m`
  const hours = Math.floor(minutes / 60)
  const rest = minutes % 60
  return rest === 0 ? `${hours}h` : `${hours}h ${rest}m`
}

File: backend/app/models/progress.py
"""Read models for the Progress screen.

None of these are stored. Each one is the shape of a single question the screen
asks — how much is left, how much is covered, how strong each topic is, and
what the last month of effort actually looked like — computed by
`services/progress.py` and never written back.
"""

from pydantic import BaseModel, Field

from app.models.common import Paper, PyObjectId


class ExamCountdown(BaseModel):
    """Days to an exam, and the honest version of that number."""

    date: str
    #: Calendar days from today. Negative once the exam is behind her.
    days: int
    #: Days left after off-days and the standing weekly off come out. This is
    #: the figure every pace calculation divides by.
    study_days: int


class Countdown(BaseModel):
    date: str
    prelims: ExamCountdown
    mains: ExamCountdown


class BurndownPoint(BaseModel):
    date: str
    #: Leaves not yet started as of this date. Null in the future, where there
    #: is nothing to report yet — the chart draws the actual line up to today.
    remaining: int | None = None
    #: Where the line has to be on this date to reach zero by Prelims, given
    #: the study days between here and there.
    required: float


class Burndown(BaseModel):
    date: str
    total_leaves: int
    started_leaves: int
    remaining: int
    study_days_remaining: int
    #: Leaves per study day needed from today to finish before Prelims.
    required_per_day: float
    #: Leaves per study day actually started over the recent window. Null until
    #: there is a window's worth of history to average.
    actual_per_day: float | None = None
    #: How many days the window measured, so the screen can say "over 28 days".
    actual_window_days: int
    #: Where the current pace lands. Null if nothing is moving.
    projected_finish: str | None = None
    series: list[BurndownPoint] = Field(default_factory=list)


class PaperCoverage(BaseModel):
    """One paper's three bars, as counts rather than percentages — the screen
    needs the denominator to say "41 of 96"."""

    paper: Paper
    label: str
    leaves: int
    read: int
    revised: int
    tested: int


class Coverage(BaseModel):
    date: str
    papers: list[PaperCoverage] = Field(default_factory=list)
    totals: PaperCoverage | None = None


class HeatmapCell(BaseModel):
    """One leaf, as a square. Tapping it opens the node."""

    node_id: PyObjectId
    title: str
    paper: Paper
    #: The level-1 ancestor, so the grid can be grouped under section headings.
    section: str
    #: 1-5, or null for a topic never graded. Null draws as an empty square,
    #: which is a different thing from a weak one.
    confidence: int | None = None
    started: bool = False
    next_due: str | None = None
    days_overdue: int = 0


class HeatmapSection(BaseModel):
    paper: Paper
    label: str
    section: str
    cells: list[HeatmapCell] = Field(default_factory=list)


class Heatmap(BaseModel):
    date: str
    sections: list[HeatmapSection] = Field(default_factory=list)


class EffortDay(BaseModel):
    date: str
    minutes: int = 0
    logs: int = 0
    #: A planned off-day is drawn as marked, not as a zero. The distinction is
    #: the whole point of tracking off-days instead of streaks.
    off: bool = False


class Effort(BaseModel):
    date: str
    days: list[EffortDay] = Field(default_factory=list)
    total_minutes: int = 0
    #: Averaged over study days only, so a fortnight's planned break does not
    #: read as a collapse in effort.
    average_minutes: int = 0
    study_days: int = 0
